cabal-install/Distribution/Solver/Modular/WeightedPSQ.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE DeriveFunctor, DeriveFoldable, DeriveTraversable #-}
+module Distribution.Solver.Modular.WeightedPSQ (
+    WeightedPSQ
+  , fromList
+  , toList
+  , keys
+  , weights
+  , length
+  , degree
+  , isZeroOrOne
+  , filter
+  , lookup
+  , mapWithKey
+  , mapWeightsWithKey
+  , union
+  ) where
+
+import qualified Data.Foldable as F
+import qualified Data.List as L
+import Data.Ord (comparing)
+import qualified Data.Traversable as T
+import Prelude hiding (filter, length, lookup)
+
+import Distribution.Solver.Modular.Degree
+
+-- | An association list that is sorted by weight.
+--
+-- Each element has a key ('k'), value ('v'), and weight ('w'). All operations
+-- that add elements or modify weights stably sort the elements by weight.
+newtype WeightedPSQ w k v = WeightedPSQ [(w, k, v)]
+  deriving (Eq, Show, Functor, F.Foldable, T.Traversable)
+
+-- | /O(N)/.
+filter :: (v -> Bool) -> WeightedPSQ k w v -> WeightedPSQ k w v
+filter p (WeightedPSQ xs) = WeightedPSQ (L.filter (p . triple_3) xs)
+
+-- | /O(N)/.
```

Comment:
`length` is actually no longer called.  I think it was used by --reorder-goals, which now uses `degree` to avoid traversing the whole list.  Calling `length` there had a significant impact on performance because it was called on the children of every available goal choice after the children were filtered with an expensive predicate.  I'll remove it.